### DISCUSSION TITLE
Include default global middleware for interacting with "Accept-Language" HTTP header

### DISCUSSION THREAD

**Discussion by SimonARG** [laravel/framework#52914] [Ideas] [answered]
# Rationale:

It's 2024. All projects are global, yet Laravel doesn't include any way to automatically detect client-side locale.

In my mind there's no reason to not use this as default behaviour, however I would understand shipping the middleware without registering it, or making it route-specific instead of global.

As for other methods for setting locale like URL parameters, for example: `/api/{LOCALE}/v1/`  being set to either use a function or middleware to set the locale depending on the `{LOCALE}` parameter. I just don't think that this, or any other alternative methods, are useful, when the HTTP spec gives us a standard method in the `Accept-Language` header.

And don't get me started on using `<input type='hidden' name='locale' value='LOCALE'>`...

`app/Http/Middleware/SetLocale.php`:

```php
namespace App\Http\Middleware;

use Closure;
use Illuminate\Support\Facades\App;

class SetLocale
{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next)
    {
        if ($locale = $request->header('Accept-Language')) {
            App::setLocale($locale);
        }

        return $next($request);
    }
}

```

`bootstrap/app.php`:

```php
->withMiddleware(function (Middleware $middleware) {
     $middleware->append(SetLocale::class);
})
```

**Comment by LucaRed** [comment]
It's not that simple, because the Accept-Language header may contain [multiple weighted codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#syntax) or even invalid codes (never trust client-side provided data). But I like the general idea of standardizing that as a default middleware.

**Reply by shaedrich** [reply]
One reason more to not let any developer temper with this by parsing it on their own but providing a standardized interface, that can help interpreting the weighing (see @mbabker's [reply](https://github.com/laravel/framework/discussions/52914#discussioncomment-10777015)).

**Comment by mbabker** [comment, marked as answer]
Luckily, the Symfony request already has a `getLanguages()` method to read that header, and a `getPreferredLanguage()` method to find the preferred language.  So you really only need to use the existing API that Laravel inherits to configure your application.

**Comment by ghost** [comment]
Why should all be embedded in Laravel? External libs are not ok for this for example?

I see that all people want the framework to solve their problems. To have all they need. :)

**Reply by SimonARG** [reply]
Since we're going there, let me give you my steelman of the two sides:

1. **Yes**, I do want the framework to solve my problems and have all I need. If I didn't, I wouldn't use a framework, I would spend 15 minutes writing a barebones router, 5 minutes making a folder-class-namespace structure, and then I'm off to work. In fact, *that's all that Laravel is about*, doing things *the Laravel way*. It's an opinionated framework by nature, as is its inspiration, Ruby on Rails. So, *why not have an opinion on something both as fundamental and as simple* as is the way to set your server language based on client information?

2. However, you could say that getting a locale from the front-end *actually isn't* something fundamental, since it deals with localization, which is something that I could see being categorized as "outside of the scope of the base application". And while Laravel *does* sometimes have opinions on non-fundamental things, these are usually very soft and explicitly optional, a lot of times not even enabled in a default build. Furthermore, you have to explicitly publish the localization files, which implies that in Laravel, localization is not considered first order of business.

I already gave my rationale. I believe that in current day, localization is actually something important enough to be considered fundamental to a web application, and that the HTTP spec already provides a standard header for dealing with front-end locale. I can see people, and even most people, disagreeing, since localization doesn't actually require "custom logic", you're literally just replacing strings. I speak 4 languages so I'm biased since I always try to have wide locale support.

**Reply by ghost** [reply]
Thanks for the explanation. Like I said. My needs are not everybody's needs and vice-versa. I would vote for  an external package for this that can be used on demand.

**Reply by shaedrich** [reply]
There are imho a whole bunch of things, I'd consider less necessary to be integrated into the framework than this which have been waved through without batting an eye.

**Reply by ghost** [reply]
Amin for that.

**Reply by ghost** [reply]
Ideally the docs should contain notes for things not included in the framework by default.
 If you want locale, you can use these packages: ....
 If you want relations that include more than 3 tables, you can use these packages:...
 (If you want useless overhead, you can use these packages: ....) Joke.
etc.

**Reply by mbabker** [reply]
By this logic, you're saying the framework's translation component shouldn't be baked in but require a third-party package.  Reading the `Accept-Language` header ties in very closely to handling translations in a number of cases, and is much less overhead compared to a number of the other framework features I routinely disable for some of the sites I deploy.

Granted, the tooling for this is already in the framework by way of the Symfony APIs that Laravel tries to steer you away from, but implementing it in middleware seems like it's a handful of lines of code at best.

**Reply by shaedrich** [reply]
> Ideally the docs should contain notes for things not included in the framework by default. If you want locale, you can use these packages: .... If you want relations that include more than 3 tables, you can use these packages:... (If you want useless overhead, you can use these packages: ....) Joke. etc.

I would quite like this 👍🏻

**Reply by ghost** [reply]
@mbabker I agree that many things are already in the framework that are less important than this. But a hand full of code multiplied by a certain amount can lead to big unused overhead for many projects. But laravel will never be like this. It will become bigger and bigger, slower and slower. Now with the help of octane and franken php, more and more load will be inserted into laravel (because it can handle it). The result will be exactly what we see in the computer hardware. New laptop that could fly under windows xp/ubuntu 14.04, but hangs in ubunu 24.04 or windows 11. Politics...
